### ISSUE TITLE
Still unable to discover new devices

### ISSUE THREAD

**Issue by gobigdave** [datech/node-red-contrib-amazon-echo#144]
##### SUMMARY

Recently got an Echo Dot g3. I have HA installed on a Pi, with NodeRed as an Add-On. My Pi is on my main LAN (vlan 1). My Echo Dot is on my IoT vlan (vlan 10). I have confirmed that both can talk to each other, and I successfully access the hub with:

curl http://192.168.1.xx/api/pi/lights

This works from both the LAN and IoT subnets.

When I ask Alexa to discover devices, nothing is ever found.

I have a Unifi USG, and multicast DNS and UPnP are both on.

##### ECHO DEVICES

Echo Dot 3rd generation

##### MODULE VERSION

```paste below

```

##### CONFIGURATION

```paste below
0.1.10
```

##### CONSOLE OUTPUT

```paste below
None
```

##### STEPS TO REPRODUCE

1. Create Amazon Echo Hub with all defaults
2. Create "Office Lights" echo device and link to hub
3. Deploy
4. Verify that /api/pi/lights is returning Office Lights info from LAN and IoT
5. "Alexa, discover devices" - try from Dot on the IoT and iOS app which is on the LAN.

My Office Lights device is never discovered.

##### EXPECTED RESULTS

Expected my Office Lights device to be discovered.

##### ACTUAL RESULTS

Alexa declares that no devices were found.

**Comment by jmorris644** [comment]
I have the exact same issue. And an almost identical network setup. My UPnP was not turned on so I did that. Although Alexa has been able to discover all of my other devices with it turned off. Even across VLANs.

I can also curl from the main LAN and the VLANs.

I did some additional testing:

1. I changed the port that the echo hub node was listening on to avoid a port conflict.
2. I then add a tcp input node for port 80 and connected it to a debug node. When I did the curl I saw it come in on port 80.
3. I then did a discover from the Alexa app and did not see anything coming in on port 80.
4. I then told one of my Alexa devices to discover and again nothing appeared.

My research will continue and I will report back any new findings.

@gobigdave did you get it working?

**Comment by gobigdave** [comment]
I have not got it working yet. I don’t see any traffic coming in on port 80 either, but like you, the Echo discovers other devices even on different VLANs. Only thing I can think of is something in iptables or Docker that is preventing the multicast discovery traffic from getting through to the Pi. That has to happen before the Echo knows to query http://[home_assistant_ip]/api/description.xml. I’ve been experimenting with a lot of iptables fun, but still no dice. Wonder if it’s something about the Pi OS?

**Comment by datech** [comment]
It looks like SSDP traffic is not transferred.

I just want to remember that you have to set the container to use the host network, if Node-Red is running on Docker.

**Comment by gobigdave** [comment]
I’m pretty sure HA addons are installed using host networking, but I will verify again.

**Comment by jmorris644** [comment]
> It looks like SSDP traffic is not transferred.
> 
> I just want to remind you that you have to set the container to use the host network, if Node-Red is running on Docker.

I did. I am using Docker but not HA at the moment. I just installed IOTstack. IN addition to telling the container to use host network I also had to open up ports 80 and 1880.

**Comment by gobigdave** [comment]
When running as an add on from HA, NodeRed uses host networking mode.

**Comment by Barabba11** [comment]
it would be better to write a guide about docker and put it on troubleshooting..

**Comment by jmorris644** [comment]
I have been trying some iptable stuff too.

Other people got versions of these commands to work. I have not. (yes, I changed the echo node to 8980)

```
sudo iptables -I INPUT 1 -p tcp --dport 80 -j ACCEPT
sudo iptables -A PREROUTING -t nat -i eth0 -p tcp --dport 80 -j REDIRECT --to-port 8980
```

**Comment by gobigdave** [comment]
I tried something similar, but no luck. I’m pretty sure the issue is the SSDP traffic isn’t making it properly, so discovery never makes it to the point where port 80 or 8980 is accessed.

**Comment by jmorris644** [comment]
To help eliminate some of the variables (Docker) I spun up a brand new PI with node-red. I then installed this plugin.

At first I was not able to connect to port 80 as expected. 

I then tried running node-red as sudo. That did not work. I think because I needed to start node itself with sudo. But, instead of trying that I gave node privileges to use lower port numbers by issuing:

`sudo setcap 'cap_net_bind_service=+ep'  $(eval readlink -f `which node`)`

After doing all of this the situation is the same. I can access it using the curl command in the OP first message. I can also see traffic on port 80. I still cannot discover.

So, people MUST have this working in this type of "plain jane" environment. Now I am thinking it is the USG?

Anyone with thoughts?

**Comment by datech** [comment]
Can you give a try with the following options:
docker run -it **--network host** --user 0 --name mynodered nodered/node-red-docker

**Comment by jmorris644** [comment]
> Can you give a try with the following options:
> docker run -it **--network host** --user 0 --name mynodered nodered/node-red-docker

That is how I am running it. I have also tried having the network in host mode and not in host mode. Same results both ways.

**Comment by jmorris644** [comment]
A little more information. I decided to install contrib-alexa-local on my new test pi. 

I tried it both running node-red without sudo and with sudo. Same results for both.

As you can see by the pics that are attached. Upon deployment it looks good. When I tell Alexa to discover devices, the node sees the discovery request. But Alexa never sees the device. The saga continues....

<img width="305" alt="alexa local" src="https://user-images.githubusercontent.com/8661459/94831254-1f6a2e80-03d2-11eb-9a8c-dff718679186.PNG">
<img width="321" alt="alexa local discovery" src="https://user-images.githubusercontent.com/8661459/94831257-2002c500-03d2-11eb-8dfd-3cd55fc23635.PNG">

**Comment by gobigdave** [comment]
Very interesting. I know the Alexa-local doesn’t work with the newer devices, but it is interesting that it did see the discovery request. What port does Alexa-local listen on?

**Comment by jmorris644** [comment]
It changes based upon the SSDP request coming in. I think that is why it doesn't work anymore because the newer echo firmware uses port 80 only.

**Comment by gobigdave** [comment]
I thought it still used SSDP to determine which IPs to see if they respond to /api/description.xml. Then again, I can’t say it should really work. For example, I curled another device on my network at http://[discovered_plug_ip]/api/description.xml, and it would not accept. That tells me that there is some other mechanism that provides the url to receive the description.xml file.

**Comment by jmorris644** [comment]
Do you have a way to test it without the unifi network?

I do but it would be a very big hassle.

**Comment by gobigdave** [comment]
Looking around the docs for building an Alexa controlled device, and it mentions that the device must support Bluetooth to be discoverable. I tried putting my Pi into discovery mode, but still no dice.

**Comment by gobigdave** [comment]
Based on what I’m reading at https://developer.amazon.com/en-US/docs/alexa/alexa-gadgets-toolkit/steps-to-set-up-gadget.html, our problem may have nothing to do with mDNS or SSDP. It looks like devices are discovered first with Bluetooth, and that connection provides the url to query for discovery. It makes more sense that Alexa says to “put devices into pairing mode if necessary”. 

I tried setting up Bluetooth (mine was off since I didn’t think I needed it) according to: https://raspberry-projects.com/pi/pi-operating-systems/raspbian/bluetooth/bluetooth-commands.

Still playing, but the Alexa doc sure seems to point to Bluetooth being the issue. Does the NodeRed add on know how to deal with Bluetooth? I’m going to try a few more things.

**Comment by jmorris644** [comment]
I think that "gadgets" are specifically bluetooth. I have many devices that are WiFi only.

**Comment by jmorris644** [comment]
based on this discover interface it looks like TCPIP, Zigbee, and Zwave are all available.

[https://developer.amazon.com/en-US/docs/alexa/device-apis/alexa-discovery.html#response](url)

**Comment by Barabba11** [comment]
I think you better use an Echo/Alexa skill module instead of this module, in this case you let Alexa discover devices itself, and you will operate with devices mapped on internet by Alexa. Yes you will depend on internet avaliability after that, no internet no smarthome.
This module instead operate without internet and it has at all another concept and reason to be, you create devices to be discovered by Alexa and it doesn't pretend to connect to existing devices. Why you ask this? There is no reason to connect to devices by NR, few people will have devices in BT range of Raspberry.. so it vecomes already useful for tto few ones.. 
I think you need to follow another way for that pourpose

**Comment by jmorris644** [comment]
I am in the middle of creating an Alexa skill as your message came in. I must admit, it is one confusing process.

My preference is to not rely on the Internet for my home automation.

I might have to abandon Alexa because of that.

**Comment by Barabba11** [comment]
if your goal is to live without internet then you need to chose proper input devices and output ones, zigbee2mqtt devices seems a valid choice. Then let your NR manage all of them, and Alexa find "fake" devices that you named in NR by this module. We have just one significative limitation, this module simulate just a bulb, you can't say alexa to close your shuttes, you need to switch off your shutters to let her close them. We need to simulate more devices here, curtains, fans, heat systems.. but to do that we need to find their documentation, or sniff the protocols and copy them. You may help the author providing him those infos.. probably someone already did the job, and probably this is open source too..

**Comment by jmorris644** [comment]
>  Then let your NR manage all of them, and Alexa find "fake" devices that you named in NR by this module. 

Agreed. But it is this module that we cannot get to function. As you state, it would only be a partial solution.

For other solutions I use a combination of keypads and 433MhZ and WiFi sensors and switches.

**Comment by jmorris644** [comment]
@gobigdave Dave, I am curious. What is your interest in this stuff?

**Comment by gobigdave** [comment]
I’m starting to connect several MQTT devices that I can control with NodeRed and Home Assistant. I’m doing everything locally, but I wanted to experiment with using Alexa as well. Mostly, I don’t want to be dependent on the cloud for normal operation, but I’m ok with using Alexa to have voice. However, I have zero devices that require the cloud to operate outside of Alexa.

I’m pretty techie, so I’m tempted to setup Haaska, but it sure would have been a lot easier if this NodeRed palette would just work like it seems to for many.

**Comment by gobigdave** [comment]
Note, most of my MQTT devices have zero access to ANYTHING (local or web) except my MQTT broker and NTP.

**Comment by jmorris644** [comment]
@gobigdave I sent you an email

**Comment by Barabba11** [comment]
Have you understood how this module works? It has nothing to do with your devices, it's just for simulate a device to Alexa. When Alexa recognise them creates function to turn on your mqtt device

**Comment by gobigdave** [comment]
Yes, I understand what it is supposed to do. However, Alexa never discovers the devices when asked. Any device I setup in NodeRed properly responds to http://[rpi_ip]/api/description.xml, but when I do “Alexa, discover devices”, she finds nothing. 

I understand this palette has nothing to do with MQTT, but it will allow me to control an MQTT device with Alexa if I could get it to discover the “Amazon Echo Device” I setup to receive the commands.

**Comment by Barabba11** [comment]
you simply have a common network problem, there are different people here who asked and find solutions, I suggest you to read all topics

**Comment by gobigdave** [comment]
I did read all the topics before posting here. Tried changing ports, removing firewalls, etc. No dice. I will read again.

**Comment by jmorris644** [comment]
I too understand what it is used for. And I have the exact same issue. I also have looked at all of the posts and have tried man of them to no avail.

**Comment by gobigdave** [comment]
I did find this. If the cause, then it will be a problem if you also use Home Assistant since everything runs in Docker. I would have to remove Home Assistant to be able to shut off all the containers because of the supervisor.

———-

just to share my 6 hours troubleshooting session's result.
I went through most of your suggestions and I actually learned a lot in the process.

I had another docker (zigbee2mqtt) running with --net=host.

If nodered started AFTER zigbee2mqtt it was not working; it was not receiving SSDP messages sent to 239.255.255.250.

If nodered runs by itself with no other docker and it's started when all other dockers are already off it WORKS

thanks for this product datech

**Comment by jmorris644** [comment]
I am not running HA or igbee2mqtt. It is good news that you actually got it working in Docker though. Do you have network=host set for NR?

**Comment by gobigdave** [comment]
I’m running Home Assistant in supervisor mode. HA allows for add ons that are all deployed as containers (as is Home Assistant itself). From what I can see looking around Docker, everything runs in network host mode. I also tried this with HassOS which either takes over the PI or you run it in a VM. That didn’t work either.

My above comment seems to indicate that NodeRed has to run by itself. Otherwise, another container could eat the discovery traffic from Alexa.

**Comment by jmorris644** [comment]
Ok, I bit the bullet and booted up a brand new pi instance. I installed NR standalone and still no joy.

Both the PI and my phone using Alexa App are on the same WiFi network, FWIW. So same VLAN.

**Comment by 3ative** [comment]
I have Home Assistant, with Node-Red install on a VM on my unRaid server. 
Devices are still discoverable...

I made a quick video to show adding devices in Node-Red:
https://youtu.be/B6mYkJoupLg 

If you would like to see how I install Alexa in my Home Assistant, I made a Blooger post of tutorial videos and you can find that here: 
https://3ative.blogspot.com/2020/08/our-guide-to-integrating-alexa-in-home.html

**Comment by jmorris644** [comment]
> I have Home Assistant, with Node-Red install on a VM on my unRaid server.
> Devices are still discoverable...
> 
> I made a quick video to show adding devices in Node-Red:
> https://youtu.be/B6mYkJoupLg
> 
> If you would like to see how I install Alexa in my Home Assistant, I made a Blooger post of tutorial videos and you can find that here:
> https://3ative.blogspot.com/2020/08/our-guide-to-integrating-alexa-in-home.html

Are you using port 80 for the Echo Hub?

**Comment by 3ative** [comment]
> Are you using port 80 for the Echo Hub?
> 
> Also, it is hard to see in the video. What version of echo are you using?

Yes Port 80 for the Hub, Like I said in the tut video, everything I use is default setup(s)

If by 'Echo' you mean the Echo Dot on my desk, that's a version two. However, I have a V3 (The one with the clock) and that works too/same.

**Comment by jmorris644** [comment]
This is so frustrating. Because of the trouble we have been having with NR under Docker  I spun up a new pi. All that is on the pi is NodeRed and this contrib.

I have also tried starting node red with sudo.

Still not working..

I am at a loss.

**Comment by jmorris644** [comment]
Maybe the new version of node-red broke something? I am on 1.1.3

**Comment by 3ative** [comment]
I don't think it's the NR version, I'm on v1.1.3 too
![image](https://user-images.githubusercontent.com/51385971/95395505-b402ee00-08f6-11eb-8d56-abb534ff1f41.png)

**Comment by Barabba11** [comment]
How you installed node red? With root permissions?

**Comment by 3ative** [comment]
@Barabba11 
If you're asking me, I've already addressed that question. 😉

**Comment by 3ative** [comment]
Here's my YouTube playlist for my Alexa stuff:
https://www.youtube.com/playlist?list=PLWRTMby105biP9gE08iGLab1FP6n7_mka

**Comment by 3ative** [comment]
FYI - I just updated everything on my installation:
```
Supervisor to 247
HA to 0.116.0
Latest Node-Red
```
_Tried the test/demo again (from the video I recently made) no change, devices get discovered ok, etc._

I see many ppl complaining they can't get some HA/NR functions to work when installed in a Docker.
Like I said, I have HA installed in an UnRaid VM. I used this guide: https://community.home-assistant.io/t/hassio-on-unraid/59959

Everything else I have installed is either through the HA '**Add-on Store**' or **HACS**.

**Comment by gobigdave** [comment]
I have an NUC on the way. I plan on installing in a VM according to the HA instructions, and we will see if that works.

**Comment by 3ative** [comment]
@gobigdave 
It should sort most of problems you're facing.
- I've had HA on a VM from day one and I've very few problems over the years and the ones I do have been easy to fix.

**Comment by jmorris644** [comment]
Ok, to try and get this working here is what I have tried now. Still not working....

1. I set up a new pi with the lite image
2. I installed node red without sudo. When I did this I got two unable-to-write errors. One for `/usr/lib/node_modules` and the other for `/usr/bin`. I changes the security on both so it would install.
3. I then installed this contrib.
4. It did not work.
5. I then ran the two ip table commands recommended in the troubleshooting section

```
sudo iptables -I INPUT 1 -p tcp --dport 80 -j ACCEPT
sudo iptables -A PREROUTING -t nat -i eth0 -p tcp --dport 80 -j REDIRECT --to-port 8080
```

6. I then changed the  port in the echo hub to 8080.

Still not discoverable.... I am out of thoughts to try.

**Comment by jmorris644** [comment]
I just installed node-red and this contrib on a windows machine and it WORKED!!!! So, it is something specific with the pi installations that is preventing it from working.

**Comment by gobigdave** [comment]
I wonder what the deal with the Pi is. I tried everything firewall related. Possible that the wired network was a problem? I didn’t get a chance to try the wireless network.

**Comment by 3ative** [comment]
@jmorris644 @gobigdave 

It's Defo a Pi thing and how it uses the network - I still think it's a Port 80 issue. 

Like I said, I've never had any issues (that weren't easy to fix) on my rig
 - Apart from HA '_Breaking Changes_' but, that's _par for the course_.

**Comment by jmorris644** [comment]
Further testing:

Thinking the problem might be Debian I installed Ubuntu on a Pi. Same issue. Not discoverable...

At least Ubuntu provided the hub node with an error initially on port 80. Which I never got with Debian.

I really don't want to run home network stuff on Windows. So I will keep trying...

**Comment by Barabba11** [comment]
why not use a raspbian like all of us? You need absolutely to install NR from the "no root" user and map the 80 port from a >1024 port. Avoid use 2 network interfaces.

**Comment by jmorris644** [comment]
> why not use a raspbian like all of us? You need absolutely to install NR from the "no root" user and map the 80 port from a >1024 port. Avoid use 2 network interfaces.

Raspbien (Now Pi OS) is Debian. Same thing. Sorry for the confusion. Not working for me.

**Comment by Barabba11** [comment]
how you can have response on both port 80 and port 8080?? It should be impossible. 
Node red should be installed without root priviledges, as I told, and it sould not be allowed to use port 80. If it does there is something wrong

**Comment by jmorris644** [comment]
I understand and have followed those instructions. I can read the xml file through port 8080 when I do not use Sudo, and when I try Sudo I can read it through port 80. So normal http commands ae getting through. I think that there is an issue with IGMP and SSDP. Your contrib IS working fine on a Windows NR that I spun up.

**Comment by Barabba11** [comment]
Again, I tell you, you don't have to sun node red from root, you don't have to install it from root.
PLease reinstall everything again and use Raspbian. Thank you for your efforts trying to solve this issue yourself, to all of us it works

**Comment by gobigdave** [comment]
I received a new NUC today, install Ubuntu server and setup a VirtualBox VM for a clean install of HassOS. I restored the snapshot of my original Pi installation so I got all my config back. HA works perfectly -- much snappier now as well. Again, this is HassOS, so it is a black box in the VM. I have zero control over what happens in the VM. HassOS is in full control.

Still does not discover new devices. Needless to say, I'm disappointed.

**Comment by 3ative** [comment]
@gobigdave 
I'm shocked, from what I can tell, it sounds like you have [_pretty much_] the same setup as me now - and mine works without issue.

When state _"NodeRed is installed as an AddOn."_ - is that within Ha or a separate install?

**Comment by gobigdave** [comment]
NodeRed is installed in HA from the AddOn store. I’m tempted to try it elsewhere just to see if it works.

**Comment by randytsuch** [comment]
I'm running NR v1.1.3, recently upgraded NR and node js.  Running on separate pi3b, recently moved HA to a pi4.  I'm using an io pin as an input to NR, so I left it on a dedicated pi to make that easier.
I tried running sudo node-red to start, and if I do this, the echo node says its connected, but I can't discover any devices.
With iptables commands, I could not even get the echo node connected.

But I tried running NR on a PC I have running 24/7, and discover devices works great there.  I'm going to run all my alexa related routines over there.

I've been using node-red-contrib-alexa-home-skill.  But it sometimes doesn't work (works maybe 75% of the time), and is cloud based so I'm happy I found that running on a PC works.

**Comment by gobigdave** [comment]
After trying on a PC, a Pi4, and a VM on a NUC without success, I gave up. I’m trying the HA Cloud, and it worked immediately. Discovered all my new devices instantly, and it works reliably. Worth the $60/year for me.

**Comment by 3ative** [comment]
After making another NR install within Docker on my UnRaid server (as opposed to having the one within HA)
I could only get Alexa nodes to work (discover devices) if I iptable-ed port 80 to to something else - I can't remember what port I used.

But that screwed the UnRaid web interface. Lluckily I didn't make the iptable permanent and a server reboot got me back in and I gave up! LOL

Next thing I'm gonna look at is why NR in HA allows traffic on Port 80 and thus allowing Alexa nodes to communicate out.

**Comment by Barabba11** [comment]
You probably guys have lack of knownledge how protocol TCP/IP works and how to debug it. I'm reading who is trying to run NR as root (as been told here many times to don't do that) or who can't figure out how another service already running on 80.

**Comment by gobigdave** [comment]
Ok, barabba11, Whatever you say. I’ve been writing software for 30+ years, CTO of several companies. I understand how TCP/IP works. Tried running NR as root and not as root, neither worked. I know how to scan if a process is using port 80 (none were). Tried same VLAN, different VLAN — didn’t work. Tried a Pi, Ubuntu NUC, Windows PC, installed NR directly or with Docker for all — none worked. Removed all firewall and iptables rules — didn’t work. 

Bottom line is that multicast and discovery work fine for all devices except this plugin on NR for me.

You need to back off your high horse. You’re not smarter than everyone else.

**Comment by imightbelosthere** [comment]
I'm at the same stage as most of you here and I'm still trying to make my Amazon Echo Gen 2 to work with the NR instance deployed on a docker container.
I am tempted on deploying the HASS appliance on a VM and see if it behaves differently, but my guess is... I'll still find myself at the same stage as everyone else here.

My Amazon Echo can find the printer that's on a different vLAN and all, but can't seem to find the docker container with NR on the exact same subnet as it is...
So I'm a bit lost here... 

I've checked every single point raised before (running with Host Network, uid 0, multicast goes through just fine...) so I don't really know or understand what might be causing this other than... something I'm missing.

**Comment by imightbelosthere** [comment]
So in my case... for some reason... my Echo Gen 2 decided to switch over to my Guest WiFi instead of the IOT WiFi for some reason and that was why it couldn't get to the NR. That sorted, I can work with the Flow without any issue :)

**Comment by datech** [comment]
Closing this issue as it has been inactive for a long time.
If the problem persists or you have further updates, feel free to reopen it.
